src/test/peerman_tests.cpp

```diff
@@ -0,0 +1,188 @@
+#include <script/standard.h>
+#include <serialize.h>
+#include <test/util/net.h>
+#include <test/util/setup_common.h>
+#include <util/string.h>
+#include <util/system.h>
+#include <util/time.h>
+#include <timedata.h>
+#include <validation.h>
+
+#include <array>
+#include <stdint.h>
+
+#include <boost/test/unit_test.hpp>
+
+static CService ip(uint32_t i)
+{
+    struct in_addr s;
+    s.s_addr = i;
+    return CService(CNetAddr(s), Params().GetDefaultPort());
+}
+
+/** Mock the connections interface. */
+class ConnectionsInterfaceMock : public ConnectionsInterface
+{
+public:
+    bool ForNode(NodeId id, std::function<bool(CNode* pnode)> func) override { return true; };
+    using NodeFn = std::function<void(CNode*)>;
+    void ForEachNode(const NodeFn& func) override {};
+    void ForEachNode(const NodeFn& func) const override {};
+    void PushMessage(CNode* pnode, CSerializedNetMsg&& msg) override;
+    CSipHasher GetDeterministicRandomizer(uint64_t id) const override { return {0, 0}; };
+    void WakeMessageHandler() override {};
+    bool OutboundTargetReached(bool historicalBlockServingLimit) const override { return true; };
+    std::vector<CAddress> GetAddresses(size_t max_addresses,
+                                       size_t max_pct,
+                                       std::optional<Network> network) const override { return {}; };
+    std::vector<CAddress> GetAddresses(CNode& requestor, size_t max_addresses, size_t max_pct) override { return {}; };
+    bool DisconnectNode(const CNetAddr& addr) override { return true; };
+    unsigned int GetReceiveFloodSize() const override { return 0; };
+    int GetExtraFullOutboundCount() const override { return 0; };
+    int GetExtraBlockRelayCount() const override { return 0; };
+    void SetTryNewOutboundPeer(bool flag) override {};
+    bool GetTryNewOutboundPeer() const override { return true; };
+    bool GetNetworkActive() const override { return true; };
+    bool GetUseAddrmanOutgoing() const override { return true; };
+    void StartExtraBlockRelayPeers() override {};
+    bool ShouldRunInactivityChecks(const CNode& node, std::chrono::seconds now) const override { return true; };
+    bool CheckIncomingNonce(uint64_t nonce) override { return true; };
+
+    void NodeReceiveMsgBytes(CNode& node, Span<const uint8_t> msg_bytes, bool& complete) const;
+    bool ReceiveMsgFrom(CNode& node, CSerializedNetMsg& ser_msg) const;
+
+    virtual ~ConnectionsInterfaceMock() {}
+
+    /** Count of number of each message type sent */
+    std::map<std::string, uint64_t> m_message_types_sent;
+};
+
+void ConnectionsInterfaceMock::PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
+{
+    BOOST_TEST_MESSAGE(strprintf("sending message %s to peer %d", msg.m_type, pnode->GetId()));
+    m_message_types_sent[msg.m_type]++;
+}
+
+void ConnectionsInterfaceMock::NodeReceiveMsgBytes(CNode& node, Span<const uint8_t> msg_bytes, bool& complete) const
+{
+    assert(node.ReceiveMsgBytes(msg_bytes, complete));
```

Comment:
According to our coding guidelines, [assertions shouldn't have side-effects](https://github.com/bitcoin/bitcoin/blob/28653a596ab7e0811ffec3bbb7632e17d54f8e43/doc/developer-notes.md?plain=1#L725).
```suggestion
    bool success = node.ReceiveMsgBytes(msg_bytes, complete);
    assert(success);
```